File: surfaces/app/src/data/presence.ts
/** PRESENCE heartbeat (Paseo adoption, Phase 2.1).
 *
 *  Tells the daemon whether the owner is actually here and which card is on
 *  screen, so it can stay quiet about a card he is already looking at instead
 *  of buzzing his phone about his own work. The policy itself lives in
 *  daemon/presence.py + notify.should_push; this file only reports facts.
 *
 *  The one subtlety worth keeping: the heartbeat TICK never counts as user
 *  activity. `lastActivityAt` moves only when the human actually does
 *  something (opens a card, sends, switches to the app), so an app left open
 *  on a desk goes stale after the daemon's freshness window and notifications
 *  resume - which is the correct behaviour, and the reason presence is not
 *  simply "is a socket connected".
 */
import { useEffect, useRef } from "react";
import { AppState, Platform } from "react-native";
import { create } from "zustand";

import { api } from "./client";
import { useConfig } from "./config";

/** Daemon counts a client as present for 3 minutes after its last real
 *  activity; beat well inside that so a few dropped beats are harmless. */
const BEAT_MS = 15_000;

/** The Henry chat reports focus like a card screen does, so the daemon can tell
 *  "he is reading the answer" from "a window is open somewhere" - the
 *  distinction notify.chat_reply needs and did not have (every Henry reply was
 *  suppressed as 'inapp' for a day). Must equal presence.CHAT in
 *  spine/comms/presence.py; no card id can collide, they all start with a date. */
export const CHAT_FOCUS = "chat";

interface PresenceState {
  focusedCard: string | null;
  lastActivityAt: number;
  /** the card currently on screen (null when the owner is elsewhere) */
  setFocusedCard: (id: string | null) => void;
  /** the human did something - the only thing that refreshes presence */
  noteActivity: () => void;
}

export const usePresence = create<PresenceState>((set) => ({
  focusedCard: null,
  lastActivityAt: Date.now(),
  setFocusedCard: (id) => set({ focusedCard: id, lastActivityAt: Date.now() }),
  noteActivity: () => set({ lastActivityAt: Date.now() }),
}));

async function beat(visible: boolean): Promise<void> {
  const { focusedCard, lastActivityAt } = usePresence.getState();
  try {
    await api.post("/presence", {
      device: Platform.OS,
      focused_card: focusedCard,
      app_visible: visible,
      // seconds, to match the daemon's time.time()
      last_activity_at: lastActivityAt / 1000,
    });
  } catch {
    // Presence is an optimisation, never a user-visible failure: offline or a
    // sleeping relay simply means the daemon assumes we are away and pushes,
    // which is the safe direction to be wrong in.
  }
}

/** Mount once in the root layout. */
export function usePresenceHeartbeat(): void {
  const visible = useRef(true);
  useEffect(() => {
    let timer: ReturnType<typeof setInterval> | null = null;
    const start = () => {
      if (timer) return;
      timer = setInterval(() => {
        if (useConfig.getState().hydrated) void beat(visible.current);
      }, BEAT_MS);
    };
    const stop = () => {
      if (timer) { clearInterval(timer); timer = null; }
    };
    start();
    void beat(true);

    const sub = AppState.addEventListener("change", (st) => {
      const active = st === "active";
      visible.current = active;
      if (active) {
        // coming back IS user activity, and the daemon should learn we are
        // here immediately rather than up to a beat later
        usePresence.getState().noteActivity();
        start();
      } else {
        // Tell the daemon we are gone BEFORE we stop beating, so a card that
        // finishes while the phone is in a pocket actually pushes.
        stop();
        void beat(false);
      }
    });
    return () => { stop(); sub.remove(); };
  }, []);
}

File: spine/comms/presence.py
# -*- coding: utf-8 -*-
"""PRESENCE - is the owner actually looking, and at WHAT? (Paseo adoption 2.1/2.2)

The rule this exists to enforce: never buzz the owner's phone about a card he
is already staring at. HelmDeck pushed on every turn end, so working on the
board meant being notified about your own work - which is how a notification
channel trains you to ignore it.

Three levels, and the distinction between the first two is the whole point:
  connected  - a socket/poll exists. Proves NOTHING; a forgotten browser tab is
               connected forever.
  present    - a client reported real USER ACTIVITY within FRESH_S. Presence is
               derived from the human, not from the transport.
  focused    - present AND the app is in the foreground AND the card it is
               showing is the card we are about to notify about.

  focused -> silent   (he is looking at it; a push would be noise)
  present -> in-app   (he is here but elsewhere; the app can surface it quietly)
  absent  -> push     (only now is a phone buzz the right instrument)

HelmDeck used to have exactly ONE push recipient by construction. Since W2d
(2026-08-28) it can have several - the phone in settings.push.fcm_token plus any
device that registered its own key in settings.push.devices (the watch). Paseo's
"pick one of N devices" problem STILL does not arise, but for a different
reason than before: this module does not pick at all. It decides WHETHER to
push; notify.recipients() then delivers to every paired device, because the
owner's wrist and his pocket are the same person and both should buzz.

Unchanged: several CLIENTS (phone, desktop, browser) can report presence, and
any one of them being focused is enough to stay silent.

State is in memory on purpose: a restarted daemon knows nothing about presence
and therefore falls back to pushing, which is the safe direction to be wrong in
(a missed push is worse than an extra one). A live client re-announces within
one heartbeat interval anyway.
"""
import threading, time

# A client heartbeats every ~15s; 3 minutes tolerates ~11 dropped beats before
# we declare it absent (the card specifies 3-min freshness).
FRESH_S = 180.0

_clients = {}          # key -> {"focused","visible","activity","device","user"}
_lock = threading.Lock()

# The Henry CHAT is a focus target like a card (2026-08-30). `focused` is only
# ever compared for equality, so a screen that is not a card needs nothing but a
# name no card id can collide with - every card id begins with its YYYYMMDD-
# filing date. Without a name for it notify.chat_reply could not ask "is he
# reading the answer right now" and fell back to "is any window open anywhere",
# which suppressed every chat push for a day (see chat_reply's header).
CHAT = "chat"

# Everything below is bounded because every field arrives from a client. The
# store is a dict keyed on client-supplied strings, so without these a buggy
# (or hostile) client that rotates its `device` on every beat would grow the
# daemon's memory without limit - and it heartbeats every 15s.
_MAX_FIELD = 200       # device / focused-card id length
_MAX_CLIENTS = 64      # far above any real fleet; prevents unbounded growth


def _clip(v, n=_MAX_FIELD):
    if v is None:
        return None
    s = str(v)
    return s[:n] if s else None


def record(user, device, focused_card=None, app_visible=True, activity_at=None):
    """One heartbeat. `key` is per user+device so a phone and a desktop are two
    independent presences, not one overwriting the other."""
    key = "%s/%s" % (_clip(user) or "?", _clip(device) or "?")
    now = time.time()
    try:
        at = float(activity_at) if activity_at is not None else now
    except (TypeError, ValueError):
        # a malformed timestamp must not 500 the heartbeat - treat it as "now"
        # and let the freshness window do its job
        at = now
    # Clamp a client clock running ahead of us: otherwise a skewed device would
    # look present forever (Paseo does the same).
    at = min(at, now)
    with _lock:
        _clients[key] = {"user": _clip(user), "device": _clip(device),
                         "focused": _clip(focused_card),
                         "visible": bool(app_visible), "activity": at}
        # drop everything long past the freshness window; it can never make a
        # notification decision again, it can only consume memory
        for k in [k for k, c in _clients.items() if now - c["activity"] > FRESH_S * 4]:
            if k != key:
                del _clients[k]
        # hard cap as the last line of defence: evict least-recently-active
        if len(_clients) > _MAX_CLIENTS:
            for k, _c in sorted(_clients.items(), key=lambda kv: kv[1]["activity"])[
                    :len(_clients) - _MAX_CLIENTS]:
                if k != key:
                    del _clients[k]
    return {"ok": True, "fresh_s": FRESH_S}


def _live(now=None):
    now = now or time.time()
    with _lock:
        return [c for c in _clients.values() if now - c["activity"] <= FRESH_S]


def plan(card_id, now=None):
    """What to do about an event on `card_id`: "silent" | "inapp" | "push"."""
    # A backgrounded client is not a place the owner can SEE an in-app cue -
    # only clients with the app actually visible can absorb "inapp" or
    # "silent". Without this filter, a client backgrounded up to FRESH_S ago
    # still counted as "live" and downgraded a push to "inapp", which then
    # showed nowhere: not on the phone (no push sent) and not in-app (nothing
    # open to show it). Measured 2026-08-24: a card kept asking questions
    # overnight and none of them buzzed the phone.
    live = [c for c in _live(now) if c["visible"]]
    if not live:
        return "push"
    if card_id and any(c["focused"] == card_id for c in live):
        return "silent"
    return "inapp"


def snapshot(now=None):
    """Debug/diagnostic view - who does the daemon think is here."""
    now = now or time.time()
    return {"fresh_s": FRESH_S,
            "clients": [{"user": c["user"], "device": c["device"],
                         "focused": c["focused"], "visible": c["visible"],
                         "idle_s": round(now - c["activity"], 1)}
                        for c in _live(now)]}


def clear():
    """Test hook."""
    with _lock:
        _clients.clear()
